### DISCUSSION TITLE
How to have multiple plans with ExtensionPay!

### DISCUSSION THREAD

**Discussion by benf2004** [Glench/ExtPay#40] [Show and tell]
I posted this in a separate comment but thought I would make a new post for greater visibility. I recommend having a decent understanding of how Extension Pay works before reading this. 

Hey everyone! I have figured out how you can add multiple plans to your extension (and confirmed with @Glench that "if it works it works"). In my case, I'm only offering two payment options (monthly & lifetime), so for me, it's not too complicated. It's a bit of a jerry-rig, but it's not too bad.

1. Create multiple extensions on [extensionpay.com](url), one for each "plan" you want to offer.
2. Do not automatically call `extpay.startBackground()`, because you're not sure which plan the user has selected. Instead, once a user has selected a plan, send a message to the background script with whatever plan they chose and then startBackground(). After the background has started, you can call `extpay.openPaymentPage()`
3. Whenever a user wants to switch plans, first you **have to** make sure they have canceled the other plan. I made a big warning popup dialog that explains their other plan does not automatically cancel. 
4. Then, after the user has canceled & confirmed their cancellation, you have to remove the extension pay variables from sync storage:
`chrome.storage.sync.remove(['extensionpay_api_key','extensionpay_installed_at', "extensionpay_user"]`. **Warning: This will immediately sign out the user of their 'paid' features.** While they can sign back in, make sure they understand by clicking continue, they will be immediately signed out. Some people may click the "buy now" button and then change their mind before they complete the purchase. That can be a hassle if you don't ensure the user knows what is happening.
5. Make sure you once again pass a message to the background with the new plan choice so you can call `extpay.startBackground()`. 
6. I also opted to store an additional variable in Chrome storage called `[extension_name]_[plan_name]`. So I retrieve the plan from chrome storage and then set `const extpay = ExtPay(plan_name)` This way, in my content & background scripts I can get the appropriate plan to show the correct features/upgrade buttons. 

If you have any questions or want some more examples, don't be afraid to reach out!

**Comment by ba32107** [comment]
Thanks @benf2004 I'm in exactly the same shoes as you (developing an extension that I'm planning to launch in the near future) and I also want to have multiple plans. This will come in handy for sure!

**Comment by surtic86** [comment]
Thanks! This is also next on my list to implement a Business Version and a Annually Plan

**Comment by BowTiedSwan** [comment]
thanks @benf2004 !! Do you have some code example for this? still figuring out how to set this up correctly, storage usage and sync etc.

**Comment by benf2004** [comment]
@BowTiedSwan Here's some snipetts

Background script:
![ray-so-export](https://user-images.githubusercontent.com/12115686/230747990-cf2d3ddf-ddb4-439f-bbcc-7f9a2ff0adbb.png)
```
chrome.runtime.onMessage.addListener(
    function(request, sender, sendResponse) {
        if (request === "is_paid") {
            chrome.storage.sync.get(['oa_plan'], function(result) {
                let plan = result.oa_plan
                const extpay = ExtPay(plan)
                extpay.getUser().then(user => {
                    const now = new Date();
                    const days_14 = 1000*60*60*24*14 // in milliseconds
                    let send;
                    if (user.paid) {
                        send = "true"
                    }
                    else if (user.trialStartedAt && (now - user.trialStartedAt) < days_14){
                        send = "true"
                    }
                    else {
                        send = "false"
                    }
                    sendResponse(send)
                })
            });
        }
        if (request === "monthly_activated"){
                var extpay = ExtPay('oa2gsheets');
                extpay.startBackground();
                sendResponse("monthly started")

        }
        if (request === "lifetime_activated") {
            var extpay = ExtPay('oa2gsheets-lifetime')
            extpay.startBackground();
            sendResponse("lifetime activated")
        }
        return true;
    }
);
```

And for a content script:
![ray-so-export (2)](https://user-images.githubusercontent.com/12115686/230748063-6480cc7d-68b8-42b7-a462-1d74670cea7e.png)
```
chrome.runtime.sendMessage('is_paid', function (response) {
        let is_paid = response;
        if (is_paid === "true") {
            main()
        }
        else {
            console.log("UNPAID")
        }
    });
```
